pandas/core/groupby/generic.py

```diff
@@ -1754,10 +1754,22 @@ def value_counts(
                     level=index_level, sort_remaining=False
                 )
 
-            if not self.as_index:
+            if self.as_index:
+                return result.__finalize__(self.obj, method="value_counts")
+            else:
                 # Convert to frame
-                result = result.reset_index(name="proportion" if normalize else "count")
-            return result.__finalize__(self.obj, method="value_counts")
+                name = "proportion" if normalize else "count"
+                index = result.index
+                columns = com.fill_missing_names(index.names)
```

Comment:
why is this needed?

Reply:
When the groupby "as" arguments are arrays or functions then the index level names come through as None. These have to converted to legitimate column headings for the frame. This is exactly what reset_index does internally.

Reply:
ok then pls make a private function out of this (on MI or index if needed) and consolidate both cases to call

Reply:
I did. It is com.fill_missing_names, and I changed reset_index to call it some time ago (My #44878)